src/linear_algebra/annihilating_polynomial.lean

```diff
@@ -0,0 +1,207 @@
+as defined in `field_theory.minpoly`.
+
+## Special case: endomorphism algebra
+
+Given an `R`-module `M` (`[add_comm_group M] [module R M]`)
+there are some common specialiazations which may be more familiar.
+* Example 1: `A = M →ₗ[R] M`, the endomorphism algebra of an `R`-module M.
+* Example 2: `A = n × n` matrices with entries in `R`.
+-/
+
+open_locale polynomial
+
+namespace polynomial
+
+section semiring
+
+variables {R A : Type*} [comm_semiring R] [semiring A] [algebra R A]
+
+variables (R)
+
+/--
+The informal notation `p(a)` stand for `polynomial.aeval a p`.
+Again informally, the annihilating ideal of `a` is
+`{ p ∈ R[X] | p(a) = 0 }`. This is an ideal in `R[X]`.
+The formal definition uses the kernel of the aeval map. -/
+noncomputable def ann_ideal (a : A) : ideal R[X] :=
+(aeval a).to_ring_hom.ker
+
+variables {R}
+
+/-- It is useful to refer to ideal membership sometimes
+ and the annihilation condition other times -/
+lemma mem_ann_ideal_iff_aeval_eq_zero (a : A) (p : R[X]) :
+  p ∈ ann_ideal R a ↔ aeval a p = 0 :=
+iff.rfl
+
+/-- relating ann_ideal to eval₂ -/
+lemma mem_of_eval₂_algebra_map_eq_zero (a : A) (p : R[X]) (h : eval₂ (algebra_map R A) a p = 0) :
+  p ∈ ann_ideal R a :=
+begin
+  apply (mem_ann_ideal_iff_aeval_eq_zero a p).2,
+  rw aeval_def,
+  apply h,
+end
+
+end semiring
+
+section field
+
+variables {𝕜 A : Type*} [field 𝕜] [ring A] [algebra 𝕜 A]
+variable (𝕜)
+
+open submodule
+
+/-- Since `𝕜[X]` is a principal ideal domain there is a polynomial `g` such that
+ `span 𝕜 {g} = ann_ideal a`. This picks some generator.
+ We prefer the monic generator of the ideal -/
+noncomputable def ann_ideal_generator (a : A) : 𝕜[X] :=
+let g := is_principal.generator $ ann_ideal 𝕜 a
+  in (C g.leading_coeff⁻¹) * g
+
+/-- ann_ideal_generator 𝕜 a is indeed a generator -/
+lemma span_singleton_ann_ideal_generator (a : A) :
+  span 𝕜[X] {ann_ideal_generator 𝕜 a} = ann_ideal 𝕜 a :=
+begin
+  by_cases (is_principal.generator $ ann_ideal 𝕜 a) = 0,
+  { rw ← is_principal.eq_bot_iff_generator_eq_zero at h,
+   simp only [ann_ideal_generator, h], simp, rw ← is_principal.eq_bot_iff_generator_eq_zero, },
+  { simp only [ann_ideal_generator, ann_ideal, alg_hom.to_ring_hom_eq_coe, ideal.submodule_span_eq],
+    rw ideal.span_singleton_mul_left_unit,
+    { exact is_principal.span_singleton_generator _ },
+    { rw [is_unit_C, is_unit_iff_ne_zero],
+      apply inv_ne_zero,
+      rw [ne.def, leading_coeff_eq_zero_iff_deg_eq_bot, degree_eq_bot],
+      rw [ann_ideal, alg_hom.to_ring_hom_eq_coe] at h,
+      apply h, }, },
+end
+
+/-- The annihilating ideal generator is a member of the annihilating ideal,
+  following submodule.generator_mem -/
+lemma ann_ideal_generator_mem (a : A) : ann_ideal_generator 𝕜 a ∈ ann_ideal 𝕜 a :=
+let I := ann_ideal 𝕜 a,
+    g := submodule.is_principal.generator I in
+  I.mul_mem_left (C g.leading_coeff⁻¹) (submodule.is_principal.generator_mem I)
+
+/-- sourced from submodule.mem_iff_eq_smul_generator -/
```

Comment:
```suggestion
```
You can add a description of the result if you wish, but there is no need to cite its source. Similarly below.